singlefactorauth/src/main/java/com/web3auth/singlefactorauth/SingleFactorAuth.kt

```
@@ -108,37 +126,100 @@ class SingleFactorAuth(
+            torusKey.getoAuthKeyData().privKey
+        } else {
+            torusKey.finalKeyData?.privKey
+        }
+
+        var decodedUserInfo: UserInfo? = null
+
+        try {
+            val jwt = decodeJwt(loginParams.idToken)
+            jwt.let {
+                decodedUserInfo = UserInfo(
+                    email = it?.getClaim("email")?.asString() ?: "",
+                    name = it?.getClaim("name")?.asString() ?: "",
+                    profileImage = it?.getClaim("picture")?.asString() ?: "",
+                    verifier = loginParams.verifier,
+                    verifierId = loginParams.verifierId,
+                    typeOfLogin = LoginType.JWT,
+                    state = TorusGenericContainer(params = mapOf())
+                )
             }
+        } catch (e: Exception) {
+            decodedUserInfo = loginParams.fallbackUserInfo
         }
 
-        val json = JSONObject()
-        if (torusSFAKey != null) {
-            json.put("privateKey", torusSFAKey.getPrivateKey())
-            json.put("publicAddress", torusSFAKey.getPublicAddress())
+        val sessionData = SessionData(
+            privateKey = privateKey.toString(),
+            publicAddress = publicAddress.toString(),
+            signatures = torusKey.sessionData,
+            userInfo = decodedUserInfo
+        )
+
+        val sessionId = SessionManager.generateRandomSessionKey()
+        sessionManager.setSessionId(sessionId)
+        sessionManager.createSession(gson.toJson(sessionData), ctx).whenComplete { result, err ->
+            if (err == null) {
+                SessionManager.saveSessionIdToStorage(result)
+                sessionManager.setSessionId(result)
+            }
         }
 
-        sessionManager.createSession(json.toString(), ctx).get()
-        return torusSFAKey
+        this.state = sessionData
+        return sessionData
+    }
+
+    private fun decodeJwt(token: String): JWT? {
+        return try {
```

throw error instead of returning null.